File: 1001-2000/1234-replace-the-substring-for-balanced-string/1234-replace-the-substring-for-balanced-string.cpp
class Solution {
public:
    int balancedString(string s) {

        int n = s.size();
        int limit = n / 4;

        vector<int> freq(128, 0);

        for(char c : s)
            freq[c]++;

        // ⭐ already balanced
        if(freq['Q'] == limit &&
           freq['W'] == limit &&
           freq['E'] == limit &&
           freq['R'] == limit)
            return 0;

        int l = 0;
        int ans = n;

        for(int r = 0; r < n; r++) {

            freq[s[r]]--;

            while(l <= r &&
                  freq['Q'] <= limit &&
                  freq['W'] <= limit &&
                  freq['E'] <= limit &&
                  freq['R'] <= limit) {

                ans = min(ans, r - l + 1);

                freq[s[l]]++;
                l++;
            }
        }

        return ans;
    }
};

File: 1001-2000/1234-replace-the-substring-for-balanced-string/1234-replace-the-substring-for-balanced-string.java
class Solution {
    public int balancedString(String s) {
        char[] charArr = s.toCharArray();
       int[] arr = new int[122];
       for(int i=0;i<charArr.length;i++){
           arr[charArr[i]]++;
       } 

       int n = s.length();

       int k = s.length()/4;
       //already balanced
       if(arr['Q']==k&&arr['W']==k&&arr['E']==k&&arr['R']==k){
          return 0;
       }

       int l =0,r=0;
       int minLength = n;
       while(r<charArr.length){
            arr[charArr[r]]--;
            //I am looking outside the window
            while(l<n&&arr['Q']<=k&&arr['W']<=k&&arr['E']<=k&&arr['R']<=k){
                 minLength = Math.min(minLength,r-l+1);
                 arr[charArr[l]]++;
                 l++;
            }

            r++;
       }

       return minLength;
    }
}